lib/puppetserver/ca/revoke_action.rb

```diff
@@ -0,0 +1,150 @@
+               'One or more comma separated certnames') do |certs|
+            parsed['certnames'] += certs
+          end
+          o.on('--config PUPPET.CONF', 'Custom path to puppet.conf') do |conf|
+            parsed['config'] = conf
+          end
+          o.on('--help', 'Displays this revoke specific help output') do |help|
+            parsed['help'] = help
+          end
+        end
+      end
+
+      def initialize(logger)
+        @logger = logger
+      end
+
+      def parse(args)
+        results = {}
+        parser = self.class.parser(results)
+
+        _, non_flags, malformed_flags, unknown_flags = Utils.parse_without_raising(parser, args)
+
+        valid = valid?(results,
+                       non_flags, malformed_flags, unknown_flags,
+                       parser.help)
+
+        # if there is an exit_code then Cli will return it early, so we only
+        # return an exit_code if there's an error
+        exit_code = valid ? nil : 1
+
+        return results, exit_code
+      end
+
+      def valid?(results, non_flags, malformed_flags, unknown_flags, usage)
+        errors = []
+
+        malformed_flags.each {|f| errors << "    Missing argument to flag `#{f}`" }
+        unknown_flags.each   {|f| errors << "    Unknown flag `#{f}`" }
+        non_flags.each       {|f| errors << "    Unknown input `#{f}`" }
+
+        results['certnames'].each do |certname|
+          if CERTNAME_BLACKLIST.include?(certname)
+            errors << "    Cannot manage cert named `#{certname}` from " +
+                      "the CLI, if needed use the HTTP API directly"
+          end
+        end
+
+        unless errors.empty?
+          @logger.err 'Error:'
+          errors.each {|e| @logger.err e }
+          @logger.err ''
+          @logger.err usage
+
+          return false
+        else
+          return true
+        end
+      end
+
+      def run(args)
+        certnames = args['certnames']
+        config = args['config']
+
+        errors = FileUtilities.validate_file_paths(config)
+        unless errors.empty?
+          @logger.err 'Error:'
+          errors.each {|error| @logger.err "    #{error}" }
+          return 1
+        end
+
+        puppet = PuppetConfig.parse(config)
+        unless puppet.errors.empty?
+          @logger.err 'Error:'
+          puppet.errors.each {|error| @logger.err "    #{error}" }
+          return 1
+        end
+
+        revoke_certs(certnames, puppet.settings)
+      end
+
+      def revoke_certs(certnames, settings)
+        url = HttpUtilities.make_ca_url(settings[:ca_server],
+                       settings[:ca_port],
+                       'certificate_status')
+
+        failed = HttpUtilities.with_connection(url, settings) do |connection|
+          certnames.reduce(false) do |has_failures, certname|
+            url.resource_name = certname
+            result = connection.put(REQUEST_BODY, url)
+
+            has_failures ||= !check_result(result, certname)
```

Comment:
Because the `||=` operator shortcuts execution `check_result` will stop being called after the first failure.

Reply:
Good catch. No, it's not desired, I think we want to consider the outcome of each revocation independent.

Reply:
How about instead of !check_result we just switch the true/falses below and rename the method to something like result_unsuccessful? with a question mark implying it will return a boolean

Reply:
I'm wondering how you will solve this.  Maybe 
```
certnames.reduce(Set[false])
```
and then a has_failures << !check_result(result, certname)
and return failed.include?(false)